packages/-ember-data/addon/setup-container.js

```diff
@@ -1,62 +1,54 @@
 import { DebugAdapter } from './-private';
+import { deprecate } from '@ember/debug';
+import { DEBUG } from '@glimmer/env';
 import Store from '@ember-data/store';
 
-function hasRegistration(application, registrationName) {
-  // fallback our ember-data tests necessary
-  // until we kill-off setupStore
-  // see https://github.com/emberjs/data/issues/6357
-  // or @ember/test-helpers kills off it's
-  // legacy support that calls our initializer with registry
-  // instead of application
-  if (typeof application.hasRegistration !== 'function') {
-    return application.has(registrationName);
+function initializeStore(application) {
+  // we can just use registerOptionsForType when we no longer
+  // support (deprecated) versions of @ember/test-helpers
+  // We're issuing a "private-api" deprecation for users of the
+  // deprecated @ember/test-helpers versions, but will keep
+  // this for as long as until 4.0 as needed
+  if (DEBUG && !application.registerOptionsForType) {
+    deprecate(
+      `Deprecated test syntax usage detected!\n\n\t` +
+        `This test relies on a deprecated test setup that is no longer supported by EmberData.` +
+        ` To resolve this you will need to be on a recent version of @ember/test-helpers` +
+        ` AND your tests must use \`setApplication()\` instead of \`setResolver()\` and` +
+        ` \`module()\` with \`setup*Test()\`instead of \`moduleFor*()\`.`,
+      false,
+      {
+        id: 'ember-data:legacy-test-helper-support',
+        until: '3.17',
+      }
+    );
+
+    application.optionsForType('serializer', { singleton: false });
```

Comment:
And then lastly, do we want `let registerOptionsForType = application.registerOptionsForType || application.optionsForType;` here still?

Reply:
No I set it up so that in production builds the legacy test support is stripped